### ISSUE TITLE
Feature request: make it possible to abort tasks not started.

### ISSUE THREAD

**Issue by lwhiteley** [ricokahler/pool#26]
Use case:
Long running processes that can be aborted by the user.

would possibly be a breaking change to the interface (not sure) but would be handy

**Comment by ricokahler** [comment]
I've considered adding in support for a feature like this and after a lot of head scratching, I've come to the conclusion that's not needed (however feel free to weight in!)

Let me explain my journey here lol (bear with me)

I would think the best way to support this kind of feature would be via the newer abort controller/[abort signal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) standard. However even with the abort controller and abort signals, promises still don't really have a built in mechanism for cancelling tasks that propagate easily.

Whatever task you're running has to support this abort controller mechanism so the end result of propagating an abort signal would look something like this:

```js
const abortController = new AbortController();

await pool({
  collection: [/* ... */],
  maxConcurrency: 10,
  signal: abortController.signal,
  task: (item, signal) => {
    const response = fetch('...', { signal });

    return // response.something ...
  },
});
```

The issue with this API is that the signal is already in the scope so making pool hold that signal is mostly irrelevant. There are some things we can do if `pool` accepted a signal such as stopping the next task to be run however this could not stop any _currently_ running tasks, which is probably what you want.

---

What I've come to realize about this whole new abort mechanism is that the rejection is how to handle and propagate cancellations. This is what the fetch API does when you cancel it via the abort controller so I think it makes sense to do it here too.

By design (i.e. following the semantics of `Promise.all`), `pool` will throw when it gets one rejection. If you have a bunch of tasks and want to cancel a whole `pool`, have your individual tasks accept an abort signal and when one tasks throws due to a cancellation, it'll bubble up and cancel your whole pool. If you want to discriminate a cancel from a failure, I would check the signal's `signal.aborted` field and re-throw with a specific error type.

Here's a `window.fetch` example of that:

```js
class AbortedError extends Error {}

async function fetchThing(id, signal) {
  try {
    const response = await fetch('...', { signal });
  } catch (e) {
    if (signal.aborted) throw new AbortedError();
    else throw e; // re-throw otherwise
  }
}

async function fetchManyThings(onCancel, onError) {
  const items = [/* ... */];
  const abortController = new AbortController();

  // imagine this was hooked up to a button or something
  onCancel(() => abortController.abort());

  try {
    return await pool({
      collection: items,
      task: id => fetchThing(id, abortController.signal),
      maxConcurrency: 10,
    });
  } catch (e) {
    if (e instanceof AbortedError) {
      // perfectly fine to ignore
    } else {
      // do some actually error handling
      onError(e);
    }
  }
}
```

Does that make sense? Do you still think a feature like this is warranted?

Let me know! Thanks for opening up the issue.

**Comment by lwhiteley** [comment]
yes this makes sense as ive thought of both scenarios as well.. 
ive also thought about making each task "cancellable" with some high order functionality .. but then it would be repetitive every time i need to do this.. 

But sometimes i understand that its better to use guides than making a feature a part of a tool set to make the API simpler and user friendly for many scenarios.

So maybe the solution here is to publish a well tested example for this scenario in the event other developers need to see this without digging through issues

**Comment by ricokahler** [comment]
Yeah that's a great idea! I'll try adding something to the docs later this week. Do you have any particular scenarios you think would be useful to include?

**Comment by lwhiteley** [comment]
The best example i can think of is uploading files or data in a file that needs to be split by chunks.

Scenario 1: 
- Uploading multiple images to a server with concurrency limit with option for user to click abort

Scenario 2:
- Uploading a single csv file where records are chunked.
- 5000 records in chunks of 100 = 50 requests
- user can cancel processing

These are very specific examples and either should be able to showcase the use case
